contracts/VestingEscrow.vy

```diff
@@ -0,0 +1,348 @@
+    """
+    self._check_sender_is_owner_or_manager()
+    # NOTE: Revoking more than once is futile
+
+    self.disabled_at = block.timestamp
+    revokable: uint256 = self._locked()
+
+    assert self.token.transfer(self.owner, revokable)
+    log RevokeUnvested(self.recipient, revokable)
+
+
+@external
+def change_owner(owner: address):
+    """
+    @notice Change contract owner.
+    @param owner Address of the new owner. Must be non-zero and
+           not same as the current owner.
+    """
+    self._check_sender_is_owner()
+    assert owner != empty(address), "zero owner address"
+
+    self.owner = owner
+    log OwnerChanged(owner)
+
+
+@external
+def change_manager(manager: address):
+    """
+    @notice Set contract manager.
+            Can update manager if it is already set.
+            Can be called only by the owner.
+    @param manager Address of the new manager
+    """
+    self._check_sender_is_owner()
+
+    self.manager = manager
+    log ManagerChanged(manager)
+
+
+@external
+def recover_erc20(token: address):
+    """
+    @notice Recover non-escrow tokens to owner
+    @param token Address of the ERC20 token to be recovered
+    """
+    self._check_sender_is_owner_or_manager()
+    assert token != self.token.address, "non-escrow token only"
+    recoverable: uint256 = ERC20(token).balanceOf(self)
+    assert ERC20(token).transfer(self.owner, recoverable)
+    log ERC20Recovered(token, recoverable)
+
+
+@external
+def update_voting_adapter(voting_adapter_addr: address):
+    """
+    @notice Set new voting_adapter_addr
+    @param voting_adapter_addr New VotingAdapter address
+    """
+    self._check_sender_is_recipient()
+    self.voting_adapter_addr = voting_adapter_addr
+    log VotingAdapterUpdated(voting_adapter_addr)
+
+
+@external
+def vote(voteId: uint256, supports: bool):
+    """
+    @notice Participate Aragon vote using all available tokens on the contract's balance
+    @param voteId Id of the vote
+    @param supports Support flag true - yea, false - nay
+    """
+    self._check_sender_is_recipient()
+    raw_call(
+        self.voting_adapter_addr,
+        _abi_encode(
+            LIDO_VOTING_CONTRACT_ADDR,
+            voteId,
+            supports,
+            method_id=method_id("vote(address,uint256,bool)"),
+        ),
+        is_delegate_call=True,
+    )
+
+
+@external
+def set_delegate():
+    """
+    @notice Delegate Snapshot voting power of all available tokens on the contract's balance
+    """
+    self._check_sender_is_recipient()
+    raw_call(
+        self.voting_adapter_addr,
+        _abi_encode(
+            SNAPSHOT_DELEGATE_CONTRACT_ADDR,
+            self.recipient,
+            method_id=method_id("set_delegate(address,address)"),
+        ),
+        is_delegate_call=True,
+    )
+
+
+@internal
+def _check_sender_is_owner_or_manager():
+    assert (
+        msg.sender == self.owner
+        or (msg.sender == self.manager and msg.sender != empty(address))
```

Comment:
Is it possible for `msg.sender` to be empty?

Reply:
Dunno. Copy pasta from https://github.com/lidofinance/mev-boost-relay-allowed-list/blob/cf5f44174c44516b6c61c9face26edabc22fb2ac/contracts/MEVBoostRelayAllowedList.vy#L273